### ISSUE TITLE
Future.onSuccess swallowing exceptions

### ISSUE THREAD

**Issue by galegofer** [vavr-io/vavr#2705]
Hi there,

I am having a kind of issue or misusage, when I do something like:

```java
Future.fromCompletableFuture(sendDocumentUploadedEvent(event))
        .onSuccess(result -> validateUpload(event, timestamps)) // throws exception
        .get();
```

if any exceptions happens inside the .onSuccess it's not able to leave the block being the Future successful, and the caller not being able to realized about the situation.

But if you do:

```java
Future.fromCompletableFuture(sendDocumentUploadedEvent(event))
        .map(result -> {
                validateUpload(event, timestamps); // throws exception
                return result; 
        )
        .get();
```

In this case the exception is not lost, and we get the exception leaving the map block code.

My question is , is this the behavior expected for the .onSuccess method?.

**Comment by danieldietrich** [comment]
Yes, this is expected behavior. Scala for example makes it more clear. There `onComplete` (resp. `onSuccess`) returns `Unit`, which is similar to `void` in Java. We return the same `Future` instance in order to allow chaining of multiple actions.

However, Scala does not completely swallow exceptions. They happen within another Thread and are printed to `System.err` I think. If you want to handle errors, you need to wrap your action in a `Try`.

We might think about exposing exceptions to the error output stream. How does Java's [`ForkJoinTask`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinTask.html) behave here?

---

Scala example:

<img width="591" alt="Future" src="https://user-images.githubusercontent.com/743833/154798629-d6ef8142-21d5-430c-a5bb-3f89fadfd32d.png">

<img width="1153" alt="Error" src="https://user-images.githubusercontent.com/743833/154798614-84d2ea1d-b39d-4376-8e36-340b1568c679.png">

**Comment by galegofer** [comment]
I think in this case we shouldn't just shallow the exception but some how log it, or make it noticeable to the developer.

I am not familiar with the ForkJoinTask behavior but with CompletableFuture.thenAccept, if you get an exception on the running thread it gets propagated.

Also I think we should update javadoc for onComplete, because right now is not stating that it may shallow the exceptions.

**Comment by danieldietrich** [comment]
@galegofer That makes sense, thanks!

**Comment by galegofer** [comment]
Thanks to you for creating this amazing library, lets's hope Vavr 1.0.0 will see the light ;)
